### PR TITLE
move, rename delete resource of editable repository

### PR DESCRIPTION
**Pull request by wouterj**
Continues https://github.com/symfony-cmf/resource-rest-bundle/pull/31 , but then with a branch in the main repo. This way, we can work on it together and get things finished fast.
## Depends on
- [x] https://github.com/symfony-cmf/resource/pull/23
- [x] https://github.com/symfony-cmf/resource-bundle/pull/31

### CODE REVIEW

**Comment by ElectricMaxxx** [comment]
Sorry i have no rights on all resource packages. Maybe i can get them as on all other CMF packages. 
Will have time on monday morning to work on. When somebody comes First just do it.

**Comment by wouterj** [comment]
> Sorry i have no rights on all resource packages. Maybe i can get them as on all other CMF packages.

You should have push access now.

**Comment by wouterj** [comment]
Pushed some commits to use the operation technique I proposed and also fix some testing.

I can't make the testing work though, as I can't find a way to update the DocumentManager. The document is actually moved, but the manager used by the test doesn't sync.<br>
Calling `$this->manager->clear()` doesn't work, as PHPCR will throw an exception because the previously known node `/cms/articles/foo` is removed. /cc @dbu @dantleech can you help out here?

**Comment by dantleech** [comment]
@WouterJ could it be that you are starting the PHPCR session in one process (creating the fixtures), moving in another (i.e. the web server process) and asserting using the first again? I have encountered that issue.

Maybe you could clear the document manage / logout of the PHPCR session before launching the browser test?

**Comment by wouterj** [comment]
Got it working by using `Session#refresh(true)` and `DM#clear()`. Not really nice DX, but it works.

This means that move + delete is now implemented. /cc @dantleech @ElectricMaxxx ready for a review

File: Controller/ResourceController.php [hunk 1]
```diff
@@ -39,11 +46,99 @@ public function __construct(
...
+
+        $path = '/'.ltrim($path, '/');
+
+        $requestContent = json_decode($request->getContent(), true);
+        if (!$requestContent) {
+            return $this->badRequestResponse('Only JSON request bodies are supported.');
+        }
+
+        foreach ($requestContent as $action) {
+            switch ($action['operation']) {
```
Comment (ElectricMaxxx) [inline Controller/ResourceController.php:96]: do we really need that operation? Is there no chance to change the path of the resource information or the target path of the request to use that as an information? to have some kind of operation key in the body is not very RESTful.

Comment (wouterj) [inline Controller/ResourceController.php:96]: Did you read http://williamdurand.fr/2014/02/14/please-do-not-patch-like-an-idiot/ ? 

> You can use whatever format you want as [description of changes], as far as its semantics is well-defined. That is why using PATCH to send updated values only is not suitable.
> 
> RFC 6902 defines a JSON document structure for expressing a sequence of operations to apply to a JSON document, suitable for use with the PATCH method. Here is how it looks like:
> 
> ```
> [
>     { "op": "test", "path": "/a/b/c", "value": "foo" },
>     { "op": "remove", "path": "/a/b/c" },
>     { "op": "add", "path": "/a/b/c", "value": [ "foo", "bar" ] },
>     { "op": "replace", "path": "/a/b/c", "value": 42 },
>     { "op": "move", "from": "/a/b/c", "path": "/a/b/d" },
>     { "op": "copy", "from": "/a/b/d", "path": "/a/b/e" }
> ]
> ```

**Comment by dbu** [comment]
could it be that the move happens on phpcr level instead of phpcr-odm document manager? or, as dan suggested, that it happens in separate sessions? if neither is the case, this sounds like a bug. though it must be a complicated case, as the straightforward `move` test is green: https://github.com/doctrine/phpcr-odm/blob/master/tests/Doctrine/Tests/ODM/PHPCR/Functional/MoveTest.php and also https://github.com/doctrine/phpcr-odm/blob/master/tests/Doctrine/Tests/ODM/PHPCR/Functional/MoveByAssignmentTest.php

**Comment by wouterj** [comment]
@dbu yeah, what dan suggested is exactly the situation here. However, I would expect `Manager#clear()` to solve the issue, but that's not the case. While clearing, it calls `JackalopeItem#getIdentifier()`, which checks the state and errors that the state is incorrect (node was removed). That's a bit strange to me (as I'm clearing it because I know the known state is invalid).

**Comment by dbu** [comment]
hm, indeed that sounds like a bug in the clear logic. i guess the
exception would need to be caught and ignored.

File: Controller/ResourceController.php (hunk 1 above)
Comment (ElectricMaxxx) [inline Controller/ResourceController.php:96]: OK i understand it now.

**Comment by dbu** [comment]
do we need to apply styleci to master and rebase this branch?

anything left to do? as the workaround for the phpcr-odm clear bug is in the tests, i guess we can merge with it. a report on phpcr-odm would be good though, or maybe a test to reproduce the situation.

**Comment by wouterj** [comment]
If @dantleech and @ElectricMaxxx agree, let's merge.

StyleCI failures are indeed not related to this PR and should be fixed in master instead afaics.

**Comment by dantleech** [comment]
+1 could we squash into two commits?

**Comment by ElectricMaxxx** [comment]
:+1:

**Comment by ElectricMaxxx** [comment]
Maybe we can merge the styleCI hints first.

**Comment by wouterj** [comment]
Squashed the commits

File: Controller/ResourceController.php [hunk 2]
```diff
@@ -29,21 +33,125 @@ class ResourceController
...
         $this->serializer = $serializer;
         $this->registry = $registry;
     }
 
-    public function resourceAction($repositoryName, $path)
+    /**
+     * Provides resource information.
+     *
+     * @param string $repositoryName
+     * $param string $path
```
Comment (dbu) [inline Controller/ResourceController.php]: `s/$param/@param/`

File: Controller/ResourceController.php [hunk 3]
```diff
@@ -29,21 +33,125 @@ class ResourceController
...
+     *     [{"operation": "move", "target": "/cms/new/id"}]
+     *
+     * Currently supported operations:
+     *
+     *  - move (options: target)
+     *
+     * changing payload properties isn't supported yet.
+     *
+     * @param string $repositoryName
+     * @param string $path
```
Comment (dbu) [inline Controller/ResourceController.php]: alignment is off, according to styleCI

**Comment by wouterj** [comment]
Updated the PR, ready to merge when the build passes.

**Comment by dbu** [comment]
great! thanks all!

**Pull request merged**